otter/convergence/selfheal.py

```diff
@@ -0,0 +1,193 @@
+        self.log = log.bind(otter_service="selfheal")
+        self.lock = kz_client.Lock("/selfheallock")
+        self.kz_client = kz_client
+        self.clock = clock
+        self.calls = []
+        timer = TimerService(
+            interval,
+            lambda: self._converge_all(config_func, interval).addErrback(
+                self.log.err, "self-heal-convergeall-err"))
+        timer.clock = clock
+        timer.setServiceParent(self)
+
+    def stopService(self):
+        """
+        Stop service by cancelling any remaining scheduled calls and releasing
+        the lock
+        """
+        super(SelfHeal, self).stopService()
+        self._cancel_scheduled_calls()
+        return self.lock.release()
+
+    def _cancel_scheduled_calls(self):
+        """
+        Cancel any remaining scheduled calls.
+
+        :return: Number of remaining active scheduled calls before cancelling
+        """
+        active = 0
+        for call in self.calls:
+            if call.active():
+                active += 1
+                call.cancel()
+        self.calls = []
+        return active
+
+    def health_check(self):
+        """
+        Return about whether this object has lock
+        """
+        d = is_lock_acquired(self.disp, self.lock)
+        return d.addCallback(lambda b: (True, {"has_lock": b}))
+
+    @inlineCallbacks
+    def _perform(self, config_func, time_range):
+        """
+        Get groups to converge and setup scheduled calls to trigger convergence
+        on each of them within time_range
+        """
+        groups = yield perform(self.disp,
+                               get_groups_to_converge(config_func))
+        active = self._cancel_scheduled_calls()
+        if active:
+            # This should never happen
+            self.log.err(RuntimeError("self-heal-calls-err"),
+                         "self-heal-calls-err", active=active)
+        if not groups:
+            returnValue(None)
+        wait_time = float(time_range) / len(groups)
+        for i, group in enumerate(groups):
+            self.calls.append(
+                self.clock.callLater(
+                    i * wait_time, perform, self.disp,
+                    check_and_trigger(group["tenantId"], group["groupId"]))
+            )
+
+    @inlineCallbacks
+    def _converge_all(self, config_func, time_range):
+        """
+        Setup convergence triggering on all groups by acquiring the lock and
+        calling _perform
+        """
+        if self.kz_client.state != KazooState.CONNECTED:
+            self.log.err(RuntimeError("self-heal-kz-state"),
+                         "self-heal-kz-state", state=self.kz_client.state)
+            returnValue(None)
+        # The reason why it checks everytime by talking to ZK is because
+        # we could've lost the lock if there where connection changes between
+        # subsequent intervals and it is not clear how state change to
+        # SUSPENDED should be handled
```

Comment:
It seems like this pattern really ought to be something like `ensure_lock_acquired`.  What you're trying to express, semantically, with the following block of code, is that `self.lock` governs mutually exclusive access to the resource represented by `self._perform`.  In order to achieve that, though, you have the actual application logic - `self._perform(config_func, time_range)` - repeated twice.  Can you factor out this logic so it can be more easily independently tested, and re-used for anything (probably any lock-protected work) which many need to repeat this pattern?